### ISSUE TITLE
Force fail CI build on the failure of test case of any integration test

### ISSUE THREAD

**Issue by budhrg** [minishift/minishift#1687]
We need to fail CI build on the failure of any test case of integration test as I feel there is no point in running build after that.

See build https://ci.centos.org/job/minishift-pr/2555/consoleFull which should be failed as soon as we get first test case failure.

**Comment by gbraad** [comment]
Can you explain more what the problem is, and what this suggestion solves?

We usually fail stages. But if an integration test fails, we would still
like to know what the results of the other tests in that stage is. If
profiles would fail, does not main that basic functionality is broken. If
proxy would fail, does not mean that other tests should not run. We still
want to know if something else got broken due to a change that got
introduced.

**Comment by budhrg** [comment]
> Can you explain more what the problem is, and what this suggestion solves?

If chosen this approach, the build will fail fast and lessen the build queue.

> We usually fail stages. But if an integration test fails, we would still
like to know what the results of the other tests in that stage is. If
profiles would fail, does not main that basic functionality is broken.

It's a choice we need to make and one way to speed up CentOS CI. However, it is the last choice I would make if [other alternatives](https://github.com/minishift/minishift/issues/1396#issuecomment-343748827) didn't work.

**Comment by hferentschik** [comment]
> It's a choice we need to make and one way to speed up CentOS CI.

How does this really speed up the build for the general case? You are "improving" the build time for an exceptional case. Hopefully, integration tests don't fail that often. If integration tests would fail on a regular basis based on pull requests, I'd argue we have another bigger problem at hand. 

When we talk about improving the build time, we are talking about improving the general/average build time. In most cases this means a build where all tests will pass. IMO, these are the things we should look into:

* Use a pre-build builder image. We spend 4 to 5 minutes installing bits and pieces. If we could have a custom builder image, we could just boot it and start our actual build targets. The latest dependencies, Golang, Ruby, etc should all be installed. @budhrg can we have custom images?
* Have a build cache. Travis and Co allows defining a cache for certain file system regions. This way one can cache build dependencies. This could, for example, be the Go dependencies or the OpenShift images. Not having to go over the network all the time for the OpenShift images would probably give a considerable speedup. 
* At some stage even with all optimisations we will reach a point though where it is not feasable anymore to run all tests on each pull request and then one has to partition them as smartly as possible. All tests should then run at a nightly basis. This is something which imo is long overdue. I think we should once and for all introduce the nightly build.

**Comment by agajdosi** [comment]
@budhrg For purpose of failing integration tests ASAP we can probably just add `--StopOnFailure=true` to the `GODOG_OPTS`. More here: https://godoc.org/github.com/DATA-DOG/godog#Options.

**Comment by gbraad** [comment]
@hferentschik we can not have images provided to the nodes, but we can use the same approach to create a builder image for the actual compile process. The install is slow, due to the actual pull of packages and the extract process. I believe having an builder image will still reduce this time, and will allow us to also have consistent build as we can guarantee the version of packages used. I think we can discuss this further on #1279.

I do not believe stopping the Integration tests is all that helpful... all changes need to follow the same tests at all times.

For local use, sure... use `--StopOnFailure=true`... but in CI, this does not sound like a good idea.

**Comment by gbraad** [comment]
I agree we should introduce nightly builds... as they can safeguard the quality in a more long-term approach. I would still like to run as much (all) coverage on each PR...

**Comment by LalatenduMohanty** [comment]
@budhrg Failing the complete CI run after one test fails can be counter productive as others have also mentioned. 

The question is why we want to improve the build time? 
 * To get feedback whether the new PR beaks anything. So that the contributor can feel more confident about the PR. 
* Faster feedback can also improve the time required to review and merge a PR. 

Am I missing anything else?
I would suggest two things to handle this. 

1. We should brake down the existing tests in to two parts. The first part should pass to check the basic functionality. It needs to pass before review can be done for the PR.

2. The second part should be run after the review is done and the PR is approved. Also we should be able to do manual trigger the tests if required. 

3. Once the second test set passes and the PR is approved, `minibot` should merge the PR automatically.

**Comment by LalatenduMohanty** [comment]
The issue with nightly tests is , the failure of nightly tests would become a technical debt for the whole team. When we keep it to per PR the responsibility  to fix it stays with the PR owner.

**Comment by hferentschik** [comment]
> The issue with nightly tests is , the failure of nightly tests would become a technical debt for the whole team. When we keep it to per PR the responsibility to fix it stays with the PR owner.

It is just not feasible to run all tests for each test. As the test suite expands you will always have to make these choices. It is a trade-off. You are running some core tests as part of the pull request and leave a full test suite run for the nightly. In most cases things should be covered by unit tests and the core integration tests. The things we want to defer are coolstore, EAP. Failures in there will very rarely be caused by code changes.

**Comment by gbraad** [comment]
CoolStore, and EAP have always been mentioned as candidates for the Nightly tests, as the testing scope here is different. We are looking regressions in the platforms and tooling we use, but also for memory usage, duration of the tests, etc.

Integration tests are meant to test for regressions in our own code (the coverage), so this should not be broken up or prevented from running. If we break up a test, even after a process of approval, would mean the merge could not happen and the developer has to review again AFTER an approval. This means we are not following the 'Fail Early' principle.

**Comment by LalatenduMohanty** [comment]
+1 for keeping the CoolStore, and EAP in nighty run. I was talking about the individual feature testing which should be carefully thought through if we want to move some of it to nightly.

**Comment by budhrg** [comment]
I am closing this issue as I found that most of the members are not in favor of the approach to `Force fail CI build on the failure of test case of any integration test`. 

We can continue our discussion to reduce build time here https://github.com/minishift/minishift/issues/1396
